algebra/Atiyah.lean

```diff
@@ -0,0 +1,1695 @@
+          rw h_1,
+          left, intro x, exact id,
+          specialize hc h_1,
+          exact hc
+        end),
+      existsi hu,
+      intros n hn,
+      cases hn with B hn,
+      cases hn with hb hn,
+      cases hb with C hc,
+      cases C with C hc,
+      cases hc with hc1 hc2,
+      cases hc with _ hc3,
+      rw ← hc1 at hn,
+      apply hc3 n hn,
+      intros A H x hx,
+      cases A with A ha,
+      exact ⟨A, ⟨⟨A, ha⟩, rfl, H⟩, hx⟩,
+      fapply exists.intro,
+      fapply subtype.mk,
+      exact zero_ideal α,
+      exact ⟨zero_ideal.is_ideal α, λ n hn, h n $ eq_of_mem_singleton hn⟩,
+      intros A ha,
+      exfalso,
+      exact ht ⟨A.val, A, rfl, ha⟩
+    end),
+  specialize Z (λ A B C hab hbc x hx, hbc $ hab hx),
+  cases Z with m Z,
+  cases m with m hm1,
+  cases hm1 with h1 h2,
+  specialize hz m,
+  have : is_prime_ideal m,
+  constructor,
+  intro hm,
+  apply h2 0,
+  rw hm,
+  trivial,
+  intros x y hxy,
+  let X : set α := quotient_to_ideal m (principal_ideal ⟦x⟧),
+  let Y : set α := quotient_to_ideal m (principal_ideal ⟦y⟧),
+  cases classical.em (∃ n, z^nat.succ n ∈ X) with hX hX,
+  cases classical.em (∃ n, z^nat.succ n ∈ Y) with hY hY,
+  cases hX with p hp,
+  cases hp with b hb,
+  cases hY with q hq,
+  cases hq with d hd,
+  cases is_ideal.coset_rep b with c hc,
+  rw [← hc, is_ideal.mul_coset] at hb,
+  cases is_ideal.coset_rep d with e he,
+  rw [← he, is_ideal.mul_coset] at hd,
+  exfalso,
+  apply h2 (p + q + 1),
+  exact calc
+          z^nat.succ (p + q + 1)
+        = z^(nat.succ p + nat.succ q) : by simp [nat.succ_eq_add_one]
+    ... = z^nat.succ p * z^nat.succ q : pow_add _ _ _
+    ... = z^nat.succ p * (y * e - (y * e - z^nat.succ q)) : by norm_num
+    ... = z^nat.succ p * (y * e) - z^nat.succ p * (y * e - z^nat.succ q) : mul_sub _ _ _
+    ... = z^nat.succ p * (y * e) - (y * e - z^nat.succ q) * z^nat.succ p : by ac_refl
+    ... = (x * c - (x * c - z^nat.succ p)) * (y * e) - (y * e - z^nat.succ q) * z^nat.succ p : by norm_num
+    ... = (x * c) * (y * e) - (x * c - z^nat.succ p) * (y * e) - (y * e - z^nat.succ q) * z^nat.succ p : by rw sub_mul
+    ... = (x * y) * (c * e) - (x * c - z^nat.succ p) * (y * e) - (y * e - z^nat.succ q) * z^nat.succ p : by ac_refl
+    ... ∈ m : is_ideal.sub_mem (is_ideal.sub_mem (is_ideal.mul_mem hxy) (is_ideal.mul_mem $ quotient.exact hb)) (is_ideal.mul_mem $ quotient.exact hd),
+  have z1 := Z ⟨Y, _, forall_not_of_not_exists hY⟩ (quotient_to_ideal.contains _ _),
+  right,
+    apply set.mem_of_mem_of_subset _ z1,
+    apply quotient_to_ideal.is_ideal,
+    apply mem_principal_ideal,
+  have z1 := Z ⟨X, _, forall_not_of_not_exists hX⟩ (quotient_to_ideal.contains _ _),
+  left,
+    apply set.mem_of_mem_of_subset _ z1,
+    apply quotient_to_ideal.is_ideal,
+    apply mem_principal_ideal,
+  specialize hz ⟨h1, this⟩,
+  apply h2 0,
+  simpa [monoid.pow, one_mul] using hz
+end
+
+-- Proposition 1.8 end
+
```

Comment:
1.8 -- the way I usually think about this result mathematically is: it's obvious
that the nilpotents are in all prime ideals. Conversely if x is not nilpotent
then S={1,x,x^2,x^3,...} is multiplicative and doesn't contain zero, so R[1/S]
(the ring of functions on the locus "x \ne 0" of Spec(R)) 
is not the zero ring (check 1 isn't 0 in the equiv relation) and hence has a prime
ideal by an earlier result. The
pre-image of this prime ideal in R is a prime ideal of R not containing x. I mention
this because you have localisation now and if localisation doesn't need 1.8 then
you could use localisation to prove 1.8 more transparently like this. 

I realise that what I want to see in a proof is completely different to what Mario
wants to see. I don't care about the Lean code. I just want some comments, which
explain to a _mathematician_ what is going on in the proof.